src/win/fs.c

```diff
@@ -142,6 +142,57 @@ void uv_fs_init(void) {
 }
 
 
+static unsigned int fs__decode_wtf8_char(const char** input) {
+  unsigned int code_point;
+  unsigned char b = **input, b2, b3, b4;
+  if (b >= 0x00 && b <= 0x7F) {
+    code_point = b;
+  } else if (b >= 0xC2 && b <= 0xDF) {
+    assert((*input)[1]);
+    b2 = *++*input;
+    code_point = ((b & 0x1F) << 6) + (b2 & 0x3F);
+  } else if (b >= 0xE0 && b <= 0xEF) {
+    assert((*input)[1] && (*input)[2]);
+    b2 = *++*input;
+    b3 = *++*input;
+    code_point = ((b & 0x0F) << 12) + ((b2 & 0x3F) << 6) + (b3 & 0x3F);
+  } else if (b >= 0xF0 && b <= 0xF0) {
+    assert((*input)[1] && (*input)[2] && (*input)[3]);
+    b2 = *++*input;
+    b3 = *++*input;
+    b4 = *++*input;
+    code_point = ((b & 0x07) << 18) + ((b2 & 0x3F) << 12) +
+      ((b3 & 0x3F) << 6) + (b4 & 0x3F);
+  }
```

Comment:
```suggestion
  } else { /* b >= 0x80 && b <= 0xC2 */
      what happens here? (I think it's malformed utf8/wtf8, but it still should be handled)
  }
```

Reply:
Unlike the other direction where every possible valid/invalid UTF-16 sequence can be encoded as WTF-8 and a unique and roundtrippable way, there's no way to encode malformed (as opposed to invalid) WTF-8 as UTF-16. Malformed WTF-8 like this would have to come from the user side, wouldn't it? Two possible options:

1. Replace malformed sequences with replacement characters (U+FFFD); according to the Unicode standard;
2. Throw an error since invalid data was coming from user and on Windows there's no way to respect it.

If you choose option 1, the Unicode standard says that you should replace each well-formed but too high utf8 byte sequence with a single replacement character and emit one replacement character for each malformed byte sequence as soon as it includes a byte that precludes the sequence from being well-formed. However, since the user side, as opposed to the system side, is passing in bad data that cannot be respected here, raising an error seems reasonable.

Reply:
I think the best course of action here is to throw an error because unlike when the file system has invalid UTF-16, which is beyond the user's control and is allowed by most file systems (both Windows and UNIX allow arbitrary invalid data in file names), this is within the user's control and there is no reasonable way to handle it on Windows: there is simply no way to take a _malformed_ (as opposed to invalid but still well-formed WTF-8) UTF-8 string and ask for a path with that name on a UTF-16 file system—this is simply an impossibility.

Reply:
@StefanKarpinski see https://github.com/libuv/libuv/pull/2192#issuecomment-464457965. The string won't come directly from the user (who shouldn't know anything about WTF-8) but from the embedder, so we can expect valid input.

Reply:
All the more reason to error of that assumption fails :+1:

Reply:
An error would make sense if it can be handled. In this case, that assumption failing indicates a problem in the embedder's code, so an assert is more appropriate.

Reply:
libuv, as a design principle, does not abort when given bad input that could be detected, and therefore handled (thus generally only for passing in memory of the wrong type, since we can't protect the user against malloc/free mistakes)

Reply:
Let me explain my point of view in more detail.

WTF-8 is [intended to be used internally in self-contained systems](https://simonsapin.github.io/wtf-8/#intended-audience). That means external input (input originating outside the system) must not be WTF-8. Consequently, the program that uses libuv must not pass external input to libuv as-is, since it might be WTF-8. In other words, the input would either be produced by the program or come from libuv itself. So any decoding error would indicate a bug in the program or in libuv. Given that, how could the program meaningfully handle the error?

Reply:
I understand your point of view, and I'm telling you that libuv is not written with that philosophy of punishing the user with abort, but is written more like libc or the kernel, which are expected to return an error code and not kill your program. The use of `assert` (or abort) is for internal logic errors within libuv or egregious errors with memory (since neither can be safely and reliably detected). When errors can be returned to the embedding program safely (such as here), they can be reported as EINVAL.

Reply:
Well, I'm not comfortable with deviating further from the spec. We're already pushing it by exposing WTF-8 outside libuv, and I suspect that tolerating invalid input would make it more likely to be exposed to end users.

Perhaps WTF-8 isn't the right solution for libuv.

Reply:
I'm not asking you to deviate from the spec in any way. I'm simply asking that it follows the existing convention in libuv and return EINVAL as an error code when encountering input that it cannot handle. (and probably a significant improvement over the current situation, which is either to silently drop them or replaces them with U+FFFD depending on the OS version)

Reply:
See https://simonsapin.github.io/wtf-8/#decoding-wtf-8, for instance. 

Can we get another opinion here? I don't insist that all specs be followed to the letter, but if we're having a disagreement about a fundamental design decision in WTF-8, then it's an indication that libuv might not be the intended audience.

Reply:
Sure, but currently your implementation also is a violation of 6.3 (“This conversion never fails” or asserts, even on ill formed input). Elsewhere, the document gives 3 options for handling ill formed WTF-8 (see the section where it defines it): Reject it, ignore it, or replace with the replacement character. We’ve previously done both 2 and 3. Any of the three are likely acceptable behaviors here, so I lean towards 1. 

All you aren’t allowed to do is abort on verifiable user input errors, as that violates both the spec you just linked and the general design principles of libuv in its role as a systems library.

Reply:
This is pretty simple and has nothing to do with WTF-8:  it is unacceptable for a library like libuv (or any C library for that matter) to call `abort`. Ever. C libraries _must_ indicate errors with error codes. That's all @vtjnash is asking you to do here. If you won't do that, then this PR simply cannot be merged as is. In that case, someone will have to take this PR and modify it to return error codes instead of calling `abort`.

Reply:
>Sure, but currently your implementation also is a violation of 6.3 (“This conversion never fails” or asserts, even on ill formed input).

That's clearly an oversight given https://github.com/SimonSapin/wtf-8/commit/c6a271dc8090f6534127abfb6ad651ef1aff9023. I've submitted an issue here: https://github.com/SimonSapin/wtf-8/issues/13

>Elsewhere, the document gives 3 options for handling ill formed WTF-8 (see the section where it defines it): Reject it, ignore it, or replace with the replacement character.

Could you link to that section?

>it is unacceptable for a C libraries to call abort

I agree with that if we consider libuv's API the system's boundary. But in that case libuv is not the intended audience for WTF-8 ("Any WTF-8 data must be converted to a Unicode encoding at the system’s boundary before being emitted.").